File: src/host/inventory.ts
import { readdir, readFile, stat } from 'node:fs/promises';
import { isAbsolute, join } from 'node:path';
import type { AgentId } from '@shared/transcript';
import type { AdapterInfo, AgentHealth, AgentInventory, AgentRuntimeInfo, InventoryFile, InventoryMcp, InventoryScope, InventorySkill, McpTransport } from '@shared/inventory';
import type { AgentExt, McpSource, RuleSource } from './agentExt';

// Read-only scan of an agent's extension points (see agentExt.ts). Pure node: no vscode import, so it runs under vitest against temp directories

export interface ScanEnv {
  home: string;
  // Workspace root (the session cwd)
  cwd: string;
  platform?: NodeJS.Platform;
  env?: NodeJS.ProcessEnv;
}

export interface ScanInput {
  agent: AgentId;
  ext?: AgentExt;
  binary: string | null;
  runtime?: AgentRuntimeInfo;
  adapter?: AdapterInfo;
  health?: AgentHealth;
}

export async function scanInventory(input: ScanInput, env: ScanEnv): Promise<AgentInventory> {
  const { agent, ext, binary, runtime, adapter, health } = input;
  const base: AgentInventory = { agent, binary, runtime, adapter, health, steer: ext?.steer ?? false, config: [], mcp: [], skills: [], rules: [], scannedAt: new Date().toISOString() };
  if (!ext) return base;
  const [config, mcp, skills, rules] = await Promise.all([
    Promise.all(ext.config.map(p => fileInfo(p, env))),
    Promise.all(ext.mcp.map(s => readMcp(s, env))),
    Promise.all(ext.skills.map(d => readSkills(d, env))),
    Promise.all(ext.rules.map(r => readRules(r, env))),
  ]);
  // Devin lists ~/.agents/skills and .agents/skills for Kimi too — one skill directory can show up once per template, but never twice for the same path
  return { ...base, config, mcp: mcp.flat(), skills: dedupe(skills.flat(), s => s.path), rules: dedupe(rules.flat(), r => r.path) };
}

// `~/x` → home; `$CONFIG/x` → XDG config home (APPDATA on Windows); relative → workspace; absolute stays
export function expandPath(template: string, env: ScanEnv): string {
  if (template.startsWith('~/')) return join(env.home, template.slice(2));
  if (template.startsWith('$CONFIG/')) return join(configHome(env), template.slice('$CONFIG/'.length));
  return isAbsolute(template) ? template : join(env.cwd, template);
}

export function scopeOf(template: string): InventoryScope {
  return template.startsWith('~/') || template.startsWith('$CONFIG/') || isAbsolute(template) ? 'user' : 'project';
}

function configHome(env: ScanEnv): string {
  const platform = env.platform ?? process.platform;
  const vars = env.env ?? process.env;
  if (platform === 'win32' && vars.APPDATA) return vars.APPDATA;
  return vars.XDG_CONFIG_HOME || join(env.home, '.config');
}

async function fileInfo(template: string, env: ScanEnv): Promise<InventoryFile> {
  const path = expandPath(template, env);
  try {
    const s = await stat(path);
    return { path, scope: scopeOf(template), exists: s.isFile(), size: s.isFile() ? s.size : undefined };
  } catch {
    return { path, scope: scopeOf(template), exists: false };
  }
}

function dedupe<T>(items: T[], key: (t: T) => string): T[] {
  const seen = new Set<string>();
  return items.filter(it => { const k = key(it); if (seen.has(k)) return false; seen.add(k); return true; });
}

// MCP declarations

async function readMcp(source: McpSource, env: ScanEnv): Promise<InventoryMcp[]> {
  const path = expandPath(source.path, env);
  let text: string;
  try { text = await readFile(path, 'utf8'); } catch { return []; }
  const scope = scopeOf(source.path);
  const entries = source.format === 'toml' ? parseTomlMcp(text) : source.format === 'opencode' ? parseOpencodeMcp(text) : parseJsonMcp(text);
  return entries.map(e => ({ ...e, source: path, scope }));
}

type McpEntry = Omit<InventoryMcp, 'source' | 'scope'>;

// { "mcpServers": { name: { command, args, env } | { url } | { transport: "sse", url } } } — Devin / Kimi / Claude-compatible .mcp.json / ~/.claude.json
export function parseJsonMcp(text: string): McpEntry[] {
  const data = parseJsonLoose(text);
  const servers = isRecord(data) && isRecord(data.mcpServers) ? data.mcpServers : undefined;
  if (!servers) return [];
  return Object.entries(servers).flatMap(([name, v]) => {
    if (!isRecord(v)) return [];
    const command = str(v.command);
    const url = str(v.url) ?? str(v.serverUrl) ?? str(v.httpUrl);
    const args = Array.isArray(v.args) ? v.args.filter((a): a is string => typeof a === 'string') : [];
    const kind = str(v.type) ?? str(v.transport);
    return [{ name, transport: transportOf(kind, command, url), target: command ? [command, ...args].join(' ') : url ?? '', enabled: v.enabled !== false && v.disabled !== true }];
  });
}

// [mcp_servers.name] tables with command / args / url / type / enabled — Grok's config.toml. Enough of TOML for these tables, nothing more
export function parseTomlMcp(text: string): McpEntry[] {
  const out: McpEntry[] = [];
  let cur: { name: string; command?: string; args: string[]; url?: string; type?: string; enabled: boolean } | undefined;
  const flush = () => { if (cur) out.push({ name: cur.name, transport: transportOf(cur.type, cur.command, cur.url), target: cur.command ? [cur.command, ...cur.args].join(' ') : cur.url ?? '', enabled: cur.enabled }); cur = undefined; };
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.trim();
    if (!line || line.startsWith('#')) continue;
    if (line.startsWith('[')) {
      flush();
      const m = /^\[\s*mcp_servers\.(?:"([^"]+)"|([^\]\s.]+))\s*\]$/.exec(line);
      if (m) cur = { name: (m[1] ?? m[2])!, args: [], enabled: true };
      continue;
    }
    if (!cur) continue;
    const kv = /^([A-Za-z_][\w-]*)\s*=\s*(.+?)\s*(?:#.*)?$/.exec(line);
    if (!kv) continue;
    const [, key, value] = kv;
    switch (key) {
      case 'command': cur.command = tomlString(value!); break;
      case 'url': cur.url = tomlString(value!); break;
      case 'type': case 'transport': cur.type = tomlString(value!); break;
      case 'enabled': cur.enabled = value!.trim() !== 'false'; break;
      case 'args': cur.args = [...value!.matchAll(/"((?:[^"\\]|\\.)*)"|'([^']*)'/g)].map(m => (m[1] ?? m[2] ?? '').replace(/\\"/g, '"')); break;
      default: break;
    }
  }
  flush();
  return out;
}

// { "mcp": { name: { type: "local", command: [...] } | { type: "remote", url } } } — OpenCode's opencode.json(c).
// `local` is a stdio child process; `remote` is streamable HTTP (our `http` bucket). A missing type is inferred from command / url
export function parseOpencodeMcp(text: string): McpEntry[] {
  const data = parseJsonLoose(text);
  const servers = isRecord(data) && isRecord(data.mcp) ? data.mcp : undefined;
  if (!servers) return [];
  return Object.entries(servers).flatMap(([name, v]) => {
    if (!isRecord(v)) return [];
    const command = Array.isArray(v.command) ? v.command.filter((a): a is string => typeof a === 'string').join(' ') : str(v.command);
    const url = str(v.url);
    const kind = str(v.type);
    const transport = kind === 'local' ? 'stdio' : kind === 'remote' ? 'http' : transportOf(undefined, command, url);
    return [{ name, transport, target: command ?? url ?? '', enabled: v.enabled !== false }];
  });
}

function tomlString(v: string): string {
  const m = /^"((?:[^"\\]|\\.)*)"|^'([^']*)'/.exec(v.trim());
  return m ? (m[1] ?? m[2] ?? '').replace(/\\"/g, '"') : v.trim();
}

function transportOf(kind: string | undefined, command?: string, url?: string): McpTransport {
  const k = kind?.toLowerCase();
  if (k === 'sse') return 'sse';
  if (k === 'stdio') return 'stdio';
  if (k === 'http' || k === 'streamable-http' || k === 'streamable_http') return 'http';
  return command ? 'stdio' : url ? 'http' : 'stdio';
}

// JSON with `//` and `/* */` comments and trailing commas (Devin's config files allow them); strict JSON is tried first
export function parseJsonLoose(text: string): unknown {
  try { return JSON.parse(text); } catch { /* fall through */ }
  let out = '';
  let i = 0;
  while (i < text.length) {
    const c = text[i]!;
    if (c === '"') {
      let j = i + 1;
      while (j < text.length && text[j] !== '"') { if (text[j] === '\\') j++; j++; }
      out += text.slice(i, j + 1);
      i = j + 1;
    } else if (c === '/' && text[i + 1] === '/') {
      while (i < text.length && text[i] !== '\n') i++;
    } else if (c === '/' && text[i + 1] === '*') {
      const end = text.indexOf('*/', i + 2);
      i = end < 0 ? text.length : end + 2;
    } else { out += c; i++; }
  }
  out = out.replace(/,(\s*[}\]])/g, '$1');
  try { return JSON.parse(out); } catch { return undefined; }
}

// Skills: <dir>/<name>/SKILL.md with a YAML-ish frontmatter carrying name / description

async function readSkills(template: string, env: ScanEnv): Promise<InventorySkill[]> {
  const dir = expandPath(template, env);
  let entries: import('node:fs').Dirent[];
  try { entries = await readdir(dir, { withFileTypes: true }); } catch { return []; }
  const scope = scopeOf(template);
  const found = await Promise.all(entries.map(async (d): Promise<InventorySkill | undefined> => {
    if (d.isDirectory() || d.isSymbolicLink()) {
      const path = join(dir, d.name, 'SKILL.md');
      try {
        const fm = parseFrontmatter(await readFile(path, 'utf8'));
        return { name: fm.name ?? d.name, description: fm.description, path, scope };
      } catch { return undefined; }
    }
    // DSH also accepts a flat <name>.md next to the bundles; it needs a frontmatter name to count
    if (d.isFile() && /\.md$/i.test(d.name)) {
      const path = join(dir, d.name);
      try {
        const fm = parseFrontmatter(await readFile(path, 'utf8'));
        return fm.name ? { name: fm.name, description: fm.description, path, scope } : undefined;
      } catch { return undefined; }
    }
    return undefined;
  }));
  return found.filter((s): s is InventorySkill => s !== undefined).sort((a, b) => a.name.localeCompare(b.name));
}

// Only `key: value` lines between the first two `---` fences; folded / literal block scalars take their following indented lines
export function parseFrontmatter(text: string): { name?: string; description?: string } {
  const m = /^---\r?\n([\s\S]*?)\r?\n---/.exec(text);
  if (!m) return {};
  const out: Record<string, string> = {};
  const lines = m[1]!.split(/\r?\n/);
  for (let i = 0; i < lines.length; i++) {
    const kv = /^([A-Za-z_][\w-]*):\s*(.*)$/.exec(lines[i]!);
    if (!kv) continue;
    let value = kv[2]!.trim();
    if (value === '>' || value === '|' || value === '>-' || value === '|-') {
      const block: string[] = [];
      while (i + 1 < lines.length && /^\s+\S/.test(lines[i + 1]!)) block.push(lines[++i]!.trim());
      value = block.join(value.startsWith('>') ? ' ' : '\n');
    }
    out[kv[1]!] = unquote(value);
  }
  return { name: out.name, description: out.description };
}

function unquote(v: string): string {
  const m = /^"((?:[^"\\]|\\.)*)"$|^'([^']*)'$/.exec(v);
  return m ? (m[1] ?? m[2] ?? '').replace(/\\"/g, '"') : v;
}

// Rules: single files, or directories of *.md / *.mdc

async function readRules(source: RuleSource, env: ScanEnv): Promise<InventoryFile[]> {
  if (!source.dir) return [await fileInfo(source.path, env)];
  const dir = expandPath(source.path, env);
  const scope = scopeOf(source.path);
  try {
    const names = (await readdir(dir, { withFileTypes: true })).filter(d => d.isFile() && /\.(md|mdc)$/i.test(d.name)).map(d => d.name).sort();
    return Promise.all(names.map(async n => { const path = join(dir, n); const s = await stat(path); return { path, scope, exists: true, size: s.size }; }));
  } catch { return []; }
}

function isRecord(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null && !Array.isArray(v);
}

function str(v: unknown): string | undefined {
  return typeof v === 'string' && v ? v : undefined;
}

File: rust/crates/acpira-host/src/inventory.rs
//! Read-only scan of an agent's extension points (mirror of src/host/inventory.ts)

use std::collections::HashSet;
use std::path::{Path, PathBuf};
use std::sync::LazyLock;

use regex::Regex;
use serde_json::{Map, Value};

use acpira_shared::inventory::*;

use crate::agent_ext::{AgentExt, McpFormat};
use crate::util::now_iso;

pub struct ScanEnv {
  pub home: String,
  pub cwd: String,
}

pub struct ScanInput {
  pub agent: String,
  pub ext: Option<&'static AgentExt>,
  pub binary: Option<String>,
  pub runtime: Option<AgentRuntimeInfo>,
  pub adapter: Option<AdapterInfo>,
  pub health: Option<AgentHealth>,
}

pub async fn scan_inventory(input: ScanInput, env: &ScanEnv) -> AgentInventory {
  let mut out = AgentInventory {
    agent: input.agent,
    binary: input.binary,
    runtime: input.runtime,
    adapter: input.adapter,
    health: input.health,
    steer: input.ext.is_some_and(|e| e.steer),
    config: vec![],
    mcp: vec![],
    skills: vec![],
    rules: vec![],
    scanned_at: now_iso(),
  };
  let Some(ext) = input.ext else { return out };
  for p in ext.config {
    out.config.push(file_info(p, env).await);
  }
  for (p, format) in ext.mcp {
    out.mcp.extend(read_mcp(p, *format, env).await);
  }
  let mut seen = HashSet::new();
  for d in ext.skills {
    for s in read_skills(d, env).await {
      if seen.insert(s.path.clone()) {
        out.skills.push(s);
      }
    }
  }
  let mut seen = HashSet::new();
  for (p, dir) in ext.rules {
    for r in read_rules(p, *dir, env).await {
      if seen.insert(r.path.clone()) {
        out.rules.push(r);
      }
    }
  }
  out
}

fn config_home(env: &ScanEnv) -> PathBuf {
  if cfg!(windows)
    && let Ok(a) = std::env::var("APPDATA")
    && !a.is_empty()
  {
    return PathBuf::from(a);
  }
  match std::env::var("XDG_CONFIG_HOME") {
    Ok(x) if !x.is_empty() => PathBuf::from(x),
    _ => Path::new(&env.home).join(".config"),
  }
}

/// `~/x` → home; `$CONFIG/x` → XDG config home; relative → workspace; absolute stays
pub fn expand_path(template: &str, env: &ScanEnv) -> String {
  let p = if let Some(rest) = template.strip_prefix("~/") {
    Path::new(&env.home).join(rest)
  } else if let Some(rest) = template.strip_prefix("$CONFIG/") {
    config_home(env).join(rest)
  } else if Path::new(template).is_absolute() {
    PathBuf::from(template)
  } else {
    Path::new(&env.cwd).join(template)
  };
  p.to_string_lossy().into_owned()
}

pub fn scope_of(template: &str) -> InventoryScope {
  if template.starts_with("~/") || template.starts_with("$CONFIG/") || Path::new(template).is_absolute() {
    InventoryScope::User
  } else {
    InventoryScope::Project
  }
}

async fn file_info(template: &str, env: &ScanEnv) -> InventoryFile {
  let path = expand_path(template, env);
  match tokio::fs::metadata(&path).await {
    Ok(m) => InventoryFile { path, scope: scope_of(template), exists: m.is_file(), size: m.is_file().then_some(m.len()) },
    Err(_) => InventoryFile { path, scope: scope_of(template), exists: false, size: None },
  }
}

struct McpEntry {
  name: String,
  transport: McpTransport,
  target: String,
  enabled: bool,
}

async fn read_mcp(template: &str, format: McpFormat, env: &ScanEnv) -> Vec<InventoryMcp> {
  let path = expand_path(template, env);
  let Ok(text) = tokio::fs::read_to_string(&path).await else { return vec![] };
  let scope = scope_of(template);
  let entries = match format {
    McpFormat::Toml => parse_toml_mcp(&text),
    McpFormat::Opencode => parse_opencode_mcp(&text),
    McpFormat::Json => parse_json_mcp(&text),
  };
  entries
    .into_iter()
    .map(|e| InventoryMcp { name: e.name, transport: e.transport, target: e.target, source: path.clone(), scope, enabled: e.enabled })
    .collect()
}

fn s(v: Option<&Value>) -> Option<String> {
  v.and_then(Value::as_str).filter(|x| !x.is_empty()).map(str::to_owned)
}

fn transport_of(kind: Option<&str>, command: Option<&str>, url: Option<&str>) -> McpTransport {
  match kind.map(str::to_lowercase).as_deref() {
    Some("sse") => McpTransport::Sse,
    Some("stdio") => McpTransport::Stdio,
    Some("http" | "streamable-http" | "streamable_http") => McpTransport::Http,
    _ => {
      if command.is_some() {
        McpTransport::Stdio
      } else if url.is_some() {
        McpTransport::Http
      } else {
        McpTransport::Stdio
      }
    }
  }
}

fn parse_json_mcp(text: &str) -> Vec<McpEntry> {
  let Some(Value::Object(data)) = parse_json_loose(text) else { return vec![] };
  let Some(servers) = data.get("mcpServers").and_then(Value::as_object) else { return vec![] };
  servers
    .iter()
    .filter_map(|(name, v)| {
      let v = v.as_object()?;
      let command = s(v.get("command"));
      let url = s(v.get("url")).or_else(|| s(v.get("serverUrl"))).or_else(|| s(v.get("httpUrl")));
      let args: Vec<String> = v
        .get("args")
        .and_then(Value::as_array)
        .map(|a| a.iter().filter_map(Value::as_str).map(str::to_owned).collect())
        .unwrap_or_default();
      let kind = s(v.get("type")).or_else(|| s(v.get("transport")));
      let target = match &command {
        Some(c) => std::iter::once(c.clone()).chain(args).collect::<Vec<_>>().join(" "),
        None => url.clone().unwrap_or_default(),
      };
      Some(McpEntry {
        name: name.clone(),
        transport: transport_of(kind.as_deref(), command.as_deref(), url.as_deref()),
        target,
        enabled: v.get("enabled") != Some(&Value::Bool(false)) && v.get("disabled") != Some(&Value::Bool(true)),
      })
    })
    .collect()
}

static TOML_HEADER: LazyLock<Regex> = LazyLock::new(|| Regex::new(r#"^\[\s*mcp_servers\.(?:"([^"]+)"|([^\]\s.]+))\s*\]$"#).unwrap());
static TOML_KV: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^([A-Za-z_][\w-]*)\s*=\s*(.+?)\s*(?:#.*)?$").unwrap());
static TOML_ARG: LazyLock<Regex> = LazyLock::new(|| Regex::new(r#""((?:[^"\\]|\\.)*)"|'([^']*)'"#).unwrap());
static TOML_STR: LazyLock<Regex> = LazyLock::new(|| Regex::new(r#"^"((?:[^"\\]|\\.)*)"|^'([^']*)'"#).unwrap());

fn toml_string(v: &str) -> String {
  match TOML_STR.captures(v.trim()) {
    Some(m) => m.get(1).or(m.get(2)).map(|x| x.as_str()).unwrap_or("").replace("\\\"", "\""),
    None => v.trim().to_owned(),
  }
}

/// [mcp_servers.name] tables: enough of TOML for these tables, nothing more
fn parse_toml_mcp(text: &str) -> Vec<McpEntry> {
  struct Cur {
    name: String,
    command: Option<String>,
    args: Vec<String>,
    url: Option<String>,
    kind: Option<String>,
    enabled: bool,
  }
  let mut out = vec![];
  let mut cur: Option<Cur> = None;
  let flush = |cur: &mut Option<Cur>, out: &mut Vec<McpEntry>| {
    if let Some(c) = cur.take() {
      let target = match &c.command {
        Some(cmd) => std::iter::once(cmd.clone()).chain(c.args.clone()).collect::<Vec<_>>().join(" "),
        None => c.url.clone().unwrap_or_default(),
      };
      out.push(McpEntry {
        transport: transport_of(c.kind.as_deref(), c.command.as_deref(), c.url.as_deref()),
        name: c.name,
        target,
        enabled: c.enabled,
      });
    }
  };
  for raw in text.split('\n') {
    let line = raw.trim();
    if line.is_empty() || line.starts_with('#') {
      continue;
    }
    if line.starts_with('[') {
      flush(&mut cur, &mut out);
      if let Some(m) = TOML_HEADER.captures(line) {
        cur = Some(Cur {
          name: m.get(1).or(m.get(2)).unwrap().as_str().to_owned(),
          command: None,
          args: vec![],
          url: None,
          kind: None,
          enabled: true,
        });
      }
      continue;
    }
    let Some(c) = cur.as_mut() else { continue };
    let Some(kv) = TOML_KV.captures(line) else { continue };
    let value = &kv[2];
    match &kv[1] {
      "command" => c.command = Some(toml_string(value)).filter(|x| !x.is_empty()),
      "url" => c.url = Some(toml_string(value)).filter(|x| !x.is_empty()),
      "type" | "transport" => c.kind = Some(toml_string(value)),
      "enabled" => c.enabled = value.trim() != "false",
      "args" => {
        c.args =
          TOML_ARG.captures_iter(value).map(|m| m.get(1).or(m.get(2)).map(|x| x.as_str()).unwrap_or("").replace("\\\"", "\"")).collect()
      }
      _ => {}
    }
  }
  flush(&mut cur, &mut out);
  out
}

fn parse_opencode_mcp(text: &str) -> Vec<McpEntry> {
  let Some(Value::Object(data)) = parse_json_loose(text) else { return vec![] };
  let Some(servers) = data.get("mcp").and_then(Value::as_object) else { return vec![] };
  servers
    .iter()
    .filter_map(|(name, v)| {
      let v = v.as_object()?;
      let command = match v.get("command") {
        Some(Value::Array(a)) => Some(a.iter().filter_map(Value::as_str).collect::<Vec<_>>().join(" ")),
        other => s(other),
      };
      let url = s(v.get("url"));
      let transport = match v.get("type").and_then(Value::as_str) {
        Some("local") => McpTransport::Stdio,
        Some("remote") => McpTransport::Http,
        _ => transport_of(None, command.as_deref(), url.as_deref()),
      };
      Some(McpEntry {
        name: name.clone(),
        transport,
        target: command.or(url).unwrap_or_default(),
        enabled: v.get("enabled") != Some(&Value::Bool(false)),
      })
    })
    .collect()
}

static TRAILING_COMMA: LazyLock<Regex> = LazyLock::new(|| Regex::new(r",(\s*[}\]])").unwrap());

/// JSON with `//` and `/* */` comments and trailing commas; strict JSON first
pub fn parse_json_loose(text: &str) -> Option<Value> {
  if let Ok(v) = serde_json::from_str(text) {
    return Some(v);
  }
  let b: Vec<char> = text.chars().collect();
  let mut out = String::with_capacity(text.len());
  let mut i = 0;
  while i < b.len() {
    let c = b[i];
    if c == '"' {
      let mut j = i + 1;
      while j < b.len() && b[j] != '"' {
        if b[j] == '\\' {
          j += 1;
        }
        j += 1;
      }
      out.extend(&b[i..(j + 1).min(b.len())]);
      i = j + 1;
    } else if c == '/' && b.get(i + 1) == Some(&'/') {
      while i < b.len() && b[i] != '\n' {
        i += 1;
      }
    } else if c == '/' && b.get(i + 1) == Some(&'*') {
      let rest: String = b[i + 2..].iter().collect();
      i = match rest.find("*/") {
        Some(end) => i + 2 + rest[..end].chars().count() + 2,
        None => b.len(),
      };
    } else {
      out.push(c);
      i += 1;
    }
  }
  serde_json::from_str(&TRAILING_COMMA.replace_all(&out, "$1")).ok()
}

async fn read_skills(template: &str, env: &ScanEnv) -> Vec<InventorySkill> {
  let dir = expand_path(template, env);
  let Ok(mut rd) = tokio::fs::read_dir(&dir).await else { return vec![] };
  let scope = scope_of(template);
  let mut found = vec![];
  while let Ok(Some(d)) = rd.next_entry().await {
    let name = d.file_name().to_string_lossy().into_owned();
    let Ok(ft) = d.file_type().await else { continue };
    if ft.is_dir() || ft.is_symlink() {
      let path = Path::new(&dir).join(&name).join("SKILL.md").to_string_lossy().into_owned();
      if let Ok(text) = tokio::fs::read_to_string(&path).await {
        let fm = parse_frontmatter(&text);
        found.push(InventorySkill { name: fm.0.unwrap_or(name), description: fm.1, path, scope });
      }
    } else if ft.is_file() && name.to_lowercase().ends_with(".md") {
      let path = Path::new(&dir).join(&name).to_string_lossy().into_owned();
      if let Ok(text) = tokio::fs::read_to_string(&path).await {
        let fm = parse_frontmatter(&text);
        if let Some(n) = fm.0 {
          found.push(InventorySkill { name: n, description: fm.1, path, scope });
        }
      }
    }
  }
  found.sort_by(|a, b| locale_compare(&a.name, &b.name));
  found
}

/// An approximation of String.prototype.localeCompare for plain names: case-insensitive first, then exact
pub fn locale_compare(a: &str, b: &str) -> std::cmp::Ordering {
  a.to_lowercase().cmp(&b.to_lowercase()).then_with(|| b.cmp(a))
}

static FRONT: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^---\r?\n((?s:.*?))\r?\n---").unwrap());
static FM_KV: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^([A-Za-z_][\w-]*):\s*(.*)$").unwrap());
static QUOTED: LazyLock<Regex> = LazyLock::new(|| Regex::new(r#"^"((?:[^"\\]|\\.)*)"$|^'([^']*)'$"#).unwrap());

/// (name, description) from `key: value` lines between the first two `---` fences
pub fn parse_frontmatter(text: &str) -> (Option<String>, Option<String>) {
  let Some(m) = FRONT.captures(text) else { return (None, None) };
  let lines: Vec<&str> = m[1].split('\n').map(|l| l.strip_suffix('\r').unwrap_or(l)).collect();
  let mut map = Map::new();
  let mut i = 0;
  while i < lines.len() {
    if let Some(kv) = FM_KV.captures(lines[i]) {
      let mut value = kv[2].trim().to_owned();
      if matches!(value.as_str(), ">" | "|" | ">-" | "|-") {
        let folded = value.starts_with('>');
        let mut block = vec![];
        while i + 1 < lines.len() && lines[i + 1].starts_with(char::is_whitespace) && !lines[i + 1].trim().is_empty() {
          i += 1;
          block.push(lines[i].trim());
        }
        value = block.join(if folded { " " } else { "\n" });
      }
      let unq = match QUOTED.captures(&value) {
        Some(q) => q.get(1).or(q.get(2)).map(|x| x.as_str()).unwrap_or("").replace("\\\"", "\""),
        None => value,
      };
      map.insert(kv[1].to_owned(), Value::from(unq));
    }
    i += 1;
  }
  (map.get("name").and_then(Value::as_str).map(str::to_owned), map.get("description").and_then(Value::as_str).map(str::to_owned))
}

async fn read_rules(template: &str, dir: bool, env: &ScanEnv) -> Vec<InventoryFile> {
  if !dir {
    return vec![file_info(template, env).await];
  }
  let path = expand_path(template, env);
  let scope = scope_of(template);
  let Ok(mut rd) = tokio::fs::read_dir(&path).await else { return vec![] };
  let mut names = vec![];
  while let Ok(Some(d)) = rd.next_entry().await {
    let n = d.file_name().to_string_lossy().into_owned();
    let lower = n.to_lowercase();
    if d.file_type().await.is_ok_and(|t| t.is_file()) && (lower.ends_with(".md") || lower.ends_with(".mdc")) {
      names.push(n);
    }
  }
  names.sort();
  let mut out = vec![];
  for n in names {
    let p = Path::new(&path).join(&n);
    let size = tokio::fs::metadata(&p).await.map(|m| m.len()).ok();
    out.push(InventoryFile { path: p.to_string_lossy().into_owned(), scope, exists: true, size });
  }
  out
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn parses_mcp_and_frontmatter() {
    let t = parse_toml_mcp(
      "[mcp_servers.fs]\ncommand = \"npx\"\nargs = [\"-y\", 'srv']\n[mcp_servers.\"web x\"]\nurl = \"https://x\"\nenabled = false\n",
    );
    assert_eq!(t.len(), 2);
    assert_eq!(t[0].target, "npx -y srv");
    assert_eq!((t[1].name.as_str(), t[1].transport, t[1].enabled), ("web x", McpTransport::Http, false));
    let j = parse_json_mcp("{ // c\n \"mcpServers\": { \"a\": { \"url\": \"u\", \"type\": \"sse\" }, } }");
    assert_eq!(j[0].transport, McpTransport::Sse);
    assert_eq!(parse_frontmatter("---\nname: x\ndescription: >\n  a\n  b\n---\n"), (Some("x".into()), Some("a b".into())));
  }
}
